File: reference-sdks/nodejs/src/media/crypto.ts
import { createCipheriv, createDecipheriv, randomBytes } from 'node:crypto'
import { MediaError } from '../core/errors.js'

/**
 * AES-128-ECB encryption used by WeChat CDN.
 * All media files are encrypted before upload and decrypted after download.
 */
export function encryptAesEcb(plaintext: Buffer, key: Buffer): Buffer {
  if (key.length !== 16) {
    throw new MediaError(`AES key must be 16 bytes, got ${key.length}`)
  }
  const cipher = createCipheriv('aes-128-ecb', key, null)
  return Buffer.concat([cipher.update(plaintext), cipher.final()])
}

export function decryptAesEcb(ciphertext: Buffer, key: Buffer): Buffer {
  if (key.length !== 16) {
    throw new MediaError(`AES key must be 16 bytes, got ${key.length}`)
  }
  const decipher = createDecipheriv('aes-128-ecb', key, null)
  return Buffer.concat([decipher.update(ciphertext), decipher.final()])
}

/**
 * Generate a random 16-byte AES key.
 */
export function generateAesKey(): Buffer {
  return randomBytes(16)
}

/**
 * Calculate the encrypted file size (AES-128-ECB with PKCS7 padding).
 * Formula: ceil((rawsize + 1) / 16) * 16
 */
export function encryptedSize(rawSize: number): number {
  return Math.ceil((rawSize + 1) / 16) * 16
}

/**
 * Decode an aes_key from the protocol.
 * Handles both formats:
 *   - Format A: base64(raw 16 bytes) → 16-byte key
 *   - Format B: base64(hex string 32 chars) → hex decode → 16-byte key
 *   - Direct hex string (32 chars) from image_item.aeskey
 */
export function decodeAesKey(encoded: string): Buffer {
  // Try direct hex first (image_item.aeskey)
  if (/^[0-9a-fA-F]{32}$/.test(encoded)) {
    return Buffer.from(encoded, 'hex')
  }

  // Try base64 decode
  const decoded = Buffer.from(encoded, 'base64')

  if (decoded.length === 16) {
    // Format A: raw 16 bytes
    return decoded
  }

  if (decoded.length === 32) {
    // Format B: 32 hex ASCII chars
    const hex = decoded.toString('ascii')
    if (/^[0-9a-fA-F]{32}$/.test(hex)) {
      return Buffer.from(hex, 'hex')
    }
  }

  throw new MediaError(
    `Cannot decode AES key: base64-decoded length is ${decoded.length} (expected 16 or 32)`,
  )
}

/**
 * Encode an AES key for the getuploadurl request (hex string format).
 */
export function encodeAesKeyHex(key: Buffer): string {
  return key.toString('hex')
}

/**
 * Encode an AES key for the sendmessage CDNMedia.aes_key field.
 * Uses base64(hex string) format for compatibility with official implementation.
 */
export function encodeAesKeyBase64(key: Buffer): string {
  return Buffer.from(key.toString('hex'), 'utf8').toString('base64')
}

File: reference-sdks/python/wechatbot/crypto.py
"""AES-128-ECB encryption for WeChat CDN media files."""

from __future__ import annotations

import base64
import binascii
import os
import re
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.primitives.padding import PKCS7

from .errors import MediaError

_HEX_32 = re.compile(r"^[0-9a-fA-F]{32}$")


def encrypt_aes_ecb(plaintext: bytes, key: bytes) -> bytes:
    """Encrypt with AES-128-ECB + PKCS7 padding."""
    if len(key) != 16:
        raise MediaError(f"AES key must be 16 bytes, got {len(key)}")
    padder = PKCS7(128).padder()
    padded = padder.update(plaintext) + padder.finalize()
    cipher = Cipher(algorithms.AES(key), modes.ECB())
    enc = cipher.encryptor()
    return enc.update(padded) + enc.finalize()


def decrypt_aes_ecb(ciphertext: bytes, key: bytes) -> bytes:
    """Decrypt AES-128-ECB and remove PKCS7 padding."""
    if len(key) != 16:
        raise MediaError(f"AES key must be 16 bytes, got {len(key)}")
    if len(ciphertext) % 16 != 0:
        raise MediaError(f"Ciphertext length {len(ciphertext)} is not a multiple of 16")
    cipher = Cipher(algorithms.AES(key), modes.ECB())
    dec = cipher.decryptor()
    padded = dec.update(ciphertext) + dec.finalize()
    unpadder = PKCS7(128).unpadder()
    return unpadder.update(padded) + unpadder.finalize()


def generate_aes_key() -> bytes:
    """Generate a random 16-byte AES key."""
    return os.urandom(16)


def encrypted_size(raw_size: int) -> int:
    """Calculate size after AES-128-ECB with PKCS7 padding."""
    return ((raw_size + 1 + 15) // 16) * 16


def decode_aes_key(encoded: str) -> bytes:
    """Decode an aes_key from the protocol.

    Handles all three formats:
      - Direct hex string (32 hex chars) — from image_item.aeskey
      - base64(raw 16 bytes) — Format A
      - base64(hex string 32 chars) — Format B
    """
    # Direct hex
    if _HEX_32.match(encoded):
        return binascii.unhexlify(encoded)

    # Base64 decode
    try:
        decoded = base64.b64decode(encoded)
    except Exception as e:
        raise MediaError(f"Cannot base64 decode aes_key: {e}") from e

    if len(decoded) == 16:
        return decoded

    if len(decoded) == 32:
        try:
            hex_str = decoded.decode("ascii")
            if _HEX_32.match(hex_str):
                return binascii.unhexlify(hex_str)
        except (UnicodeDecodeError, binascii.Error):
            pass

    raise MediaError(f"Decoded aes_key has unexpected length {len(decoded)} (want 16 or 32)")


def encode_aes_key_hex(key: bytes) -> str:
    """Encode key as hex string (for getuploadurl)."""
    return key.hex()


def encode_aes_key_base64(key: bytes) -> str:
    """Encode key as base64(hex) (for CDNMedia.aes_key)."""
    return base64.b64encode(key.hex().encode("utf-8")).decode("ascii")
